Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm

```diff
@@ -546,83 +535,101 @@ static bool needsEAGLOnMac()
 }
 #endif
 
-bool GraphicsContextGLCocoa::reshapeDisplayBufferBacking()
+bool GraphicsContextGLCocoa::reshapeDrawingBuffer()
 {
     ASSERT(!getInternalFramebufferSize().isEmpty());
-    // Reset the current backbuffer now before allocating a new one in order to slightly reduce memory pressure.
-    if (m_displayBufferBacking) {
-        m_displayBufferBacking.reset();
-        EGL_ReleaseTexImage(m_displayObj, m_displayBufferPbuffer, EGL_BACK_BUFFER);
-        EGL_DestroySurface(m_displayObj, m_displayBufferPbuffer);
-        m_displayBufferPbuffer = EGL_NO_SURFACE;
-    }
-    // Reset the future recycled buffer now, because it most likely will not be reusable at the time it will be reused.
-    auto recycledBuffer = m_swapChain.recycleBuffer();
-    if (recycledBuffer.handle)
-        EGL_DestroySurface(m_displayObj, recycledBuffer.handle);
-    recycledBuffer.surface.reset();
-    return allocateAndBindDisplayBufferBacking();
+    freeDrawingBuffers();
+    return bindNextDrawingBuffer();
 }
 
 void GraphicsContextGLCocoa::setDrawingBufferColorSpace(const DestinationColorSpace& colorSpace)
 {
-    if (m_drawingBufferColorSpace != colorSpace) {
-        m_drawingBufferColorSpace = colorSpace;
+    if (m_drawingBufferColorSpace == colorSpace)
+        return;
+    m_drawingBufferColorSpace = colorSpace;
+    if (getInternalFramebufferSize().isEmpty())
+        return;
+    if (!reshapeDrawingBuffer())
+        forceContextLost();
+}
 
-        if (!getInternalFramebufferSize().isEmpty() && !reshapeDisplayBufferBacking()) {
-            RELEASE_LOG(WebGL, "Fatal: Unable to allocate backing store of size %d x %d", getInternalFramebufferSize().width(), getInternalFramebufferSize().height());
-            forceContextLost();
-        }
-    }
+GraphicsContextGLCocoa::IOSurfacePbuffer& GraphicsContextGLCocoa::displayBuffer()
+{
+    return m_drawingBuffers[(m_currentDrawingBufferIndex - 1u) % maxReusedDrawingBuffers];
```

Comment:
If/when `m_currentDrawingBufferIndex` is 0, subtracting 1 makes it roll over to the max value, and modulo a non-power-of-2 doesn't guarantee that it's the wanted 1-step-before.
In this case, (0u-1u)%3==0, which happens to be the same as (1u-1u)%3!
If you're certain that `m_currentDrawingBufferIndex` will actually start at 1u (e.g., because `bindNextDrawingBuffer()` will always be called first?), then it's fine -- though a bit scary, so maybe a comment or `assert(m_c != 0)` would be welcome.
Otherwise you'll want to tweak the formula, e.g.: `m_drawingBuffers[(m_currentDrawingBufferIndex + maxReusedDrawingBuffers - 1u) % maxReusedDrawingBuffers]`.

Reply:
@squelart nice catch!
_[Goes to stand in a corner embarrassed for such a fundamental misunderstanding]_